React context for auth state

lucaslcode opened the pull request:
Note: Parts of storybook are probably broken in this commit

Another trial. There are probably a few different ways of doing this, but this one lets you call the functions directly on the state object like in mobx.

In components I kept using `const authContext = useAppContext(AuthContext)` but maybe I should have named it `authStore` or `authState`?

I kept user in the authState, but I would like to change this to userId and get the current user from the userCache like any user.

Some drawbacks:

- I think there is a big issue that calling functions on the state inside of `useEffect` gives a warning if you don't include `authContext` in the dependencies, but if you do, the `useEffect` is called every time and part of `authContext` changes. Not sure what the best solution is.
- Re-renders the whole app every time `authContext` changes. This is maybe not a huge deal since it should very rarely change except for login.

lucaslcode committed:
Auth state using context instead of redux

Note storybook is probably not working

darrenvong commented on app/frontend/src/features/auth/useAuthStore.ts:
```diff
@@ -0,0 +1,155 @@
+  );
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  return {
```
Yeah I'm not sure there's a good way around this tbh! Usually memoising this object (with `useMemo`) would help to keep this object stable, but it's pointless here since as soon you update one of the state, the object will have to update anyway!

Like you said in the PR comment, this shouldn't cause a lot of issue as it only changes between logging in and out, so the object should be reasonably stable anyway!

darrenvong commented on app/frontend/src/features/auth/AuthProvider.tsx:
```diff
@@ -0,0 +1,17 @@
+    throw Error("No context provided!");
+  }
+  return contextValue;
+}
```
```suggestion
const AuthContext = React.createContext<null | AuthStoreType>(null);

export function useAppContext() {
  const contextValue = useContext(AuthContext);
  if (contextValue === null) {
    throw Error("No context provided!");
  }
  return contextValue;
}
```
Presumably `AppContext` is only gonna be used with this helper hook, you could do this instead. Then everywhere else in the app, you can `useAppContext` without needing `AuthContext` from this file and so it doesn't even need to be exported.

darrenvong commented on app/frontend/src/features/auth/jail/Jail.tsx:
```diff
@@ -21,29 +20,28 @@ const useStyles = makeStyles((theme) => ({
   const [loading, setLoading] = useState(false);
   const [jailInfo, setJailInfo] = useState<null | JailInfoRes.AsObject>(null);
 
+  ///TODO: This could fire twice because of authContext dependency
```
Hmm I think we would be stuck in an infinite rendering loop because of this since `authContext` will be always updating once this effect runs. Maybe we could solve this by moving this initial check on login success so we don't have to run it here?

lucaslcode replied:
I might put actions into a different provider.

darrenvong commented:
> Parts of storybook are probably broken in this commit

Yeah, you probably will have to wrap each Story rendered component with this new AuthContext you created if any of them needs to read auth values.

lucaslcode committed:
separate authActions from authState

lucaslcode committed:
First hook test

lucaslcode committed:
move existing updateProfile tests

lucaslcode committed:
more tests

Got rid of referential equality, actually we can't have that here or it uses outdated values

lucaslcode committed:
complete useAuthStore tests

lucaslcode committed:
format

lucaslcode commented on app/frontend/src/service/__mocks__/index.ts:
```diff
@@ -51,6 +52,22 @@ export const mockedService = ({
       ]),
     getGroupChatMessages: () => Promise.resolve(messages),
   },
+  requests: {
+    listHostRequests: () =>
+      Promise.resolve([
+        {
+          hostRequestId: 1,
+          fromUserId: 1,
+          toUserId: 2,
+          status: HostRequestStatus.HOST_REQUEST_STATUS_PENDING,
+          created: { seconds: Date.now() / 1000, nanos: 0 },
+          fromDate: "2020/12/01",
+          toDate: "2020/12/06",
+          lastSeenMessageId: 0,
+          latestMessage: messages[0],
+        },
+      ]),
+  },
```
The Messages view in storybook says this mock doesn't exists so `hostRequests` is undefined. Not sure why?

darrenvong replied:
Yeah something weird is going on there, as `listHostRequests` seems to be overwritten by something that imports this mock... haven't really got much time to dig into this more yet. This is one of my gripes on using module state in anything as they make testing a pain, especially as we don't have jest here.

lucaslcode replied:
Let's just leave it - it will be refactors when I take redux out completely

lucaslcode commented on app/frontend/src/test/utils.ts:
```diff
@@ -1,13 +1,9 @@
-      user: defaultUser,
-    },
-  });
+  window.localStorage.setItem("auth.authenticated", JSON.stringify(true));
+  window.localStorage.setItem("auth.jailed", JSON.stringify(false));
+  window.localStorage.setItem("auth.user", JSON.stringify(defaultUser));
```
Is there a better way to make a "logged in user" default for tests?

darrenvong replied:
You can probably set the store directly to the user in a `beforeEach`? I pulled this out here rather than putting in `beforeEach` in setUpTests as I realised I didn't want this to be injected at the start of every tests, as some tests like the auth tests would need to start with no users first. Nothing stopping you from putting it in `beforeEach` in the test file that needs it though.

lucaslcode replied:
There's no way to set the store directly now though- is it ok to add an optional parameter to useAuthStore for testing? Or another way?

darrenvong replied:
Since you had to make a mock implementation of local storage, couldn't you set values directly against `store` exposed by `createLocalStorageMock`?

lucaslcode replied:
I don't necessarily want it in every test. What I really meant was, is there a better way than pre-setting the mocked local storage. If you think it's fine, I'll leave as is.

darrenvong replied:
Probably not, I think it's fine to have it like this and it's why I got in a helper to minimise setup boilerplate, so for test files that need an authenticated user in all tests, you can put it in that file's `beforeEach`. Otherwise call it separately in only the tests that need them.

lucaslcode committed:
logout on unauthenticated error

lucaslcode committed:
remember dependency array

darrenvong commented on app/frontend/src/features/auth/jail/Jail.tsx:
```diff
@@ -21,29 +20,27 @@ const useStyles = makeStyles((theme) => ({
       setLoading(false);
     })();
-  }, [dispatch]);
+  }, [authActions]);
```
Just tried with a new user and my browser is "running out of resource" as a result of this lol. I think this is still having the same problem as before, since `authActions` are still gonna be a new object every time state updates, so this will still be an infinite loop. I think you'll have to memoise the actions object to keep it stable for this to work.

darrenvong commented on app/frontend/src/features/auth/login/UsernameForm.tsx:
```diff
@@ -57,7 +56,7 @@ export default function UsernameForm() {
-      dispatch(authError(e.message));
+      authActions.authError(e.message);
     }
     setLoading(false);
```
I'm quite consistently getting that setState on unmounted component error on this now. That isMounted "hack" may actually be useful here...

darrenvong commented on app/frontend/src/features/auth/useAuthStore.ts:
```diff
@@ -0,0 +1,169 @@
+      }
+      setLoading(false);
+    },
+    async updateUserProfile(userData: UpdateUserProfileData) {
+      const username = user?.username;
+
+      if (!username) {
+        throw Error("User is not connected.");
+      }
+
+      await service.user.updateProfile(userData);
+
+      setUser(await service.user.getUser(username));
+    },
+    async updateHostingPreferences(preferences: HostingPreferenceData) {
+      const username = user?.username;
+
+      if (!username) {
+        throw Error("User is not connected.");
+      }
+
+      await service.user.updateHostingPreference(preferences);
+
+      setUser(await service.user.getUser(username));
+    },
```
Probably best to split this up into its own action object as it's not strictly auth related? Then you should be able to memoise the rest of the object whilst avoiding the problem of these two actions reading stale values of `user`.

lucaslcode replied:
Yes I will do that. As long as we won't need the state in any other actions that come up...

In the next PR I am going to refactor userCache to hooks and take user out of Auth completely, and just leave user ID.

darrenvong reviewed (changes requested):
Only a few things I found so far as this is quite a beasty PR! Will probably have to come back for a second round (or somebody else could have a look too), but this will be a big improvement once we get it in!

lucaslcode committed:
context review

darrenvong commented on app/frontend/src/features/auth/useAuthStore.ts:
```diff
@@ -0,0 +1,178 @@
+  const [_state, _setState] = useState<T>(
+    saved !== null ? JSON.parse(saved) : defaultValue
+  );
+  const setState = (value: T) => {
```
I think you'll have to memoise this too, otherwise it will be a different function reference each time a state update causes this hook to re-run, which will then break the `useMemo` in `useAuthStore`. Jail is still broken probably because of this :/

darrenvong replied:
Jail is still broken because of this.

darrenvong commented on app/frontend/src/features/messages/surfing/SurfingTab.tsx:
```diff
@@ -39,8 +39,10 @@ export default function SurfingTab() {
+  const hostRequests = useTypedSelector(
+    (state) => state.hostRequests.hostRequests
+  );
+  console.log(hostRequests);
```
Left over `console.log` from debugging?

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+
+  it("saves then loads a value", () => {
+    const value = "Test string";
+    renderHook(() => {
+      const [state, setState] = usePersistedState("key", "");
+      expect(state).toBe("");
+      setState(value);
+      expect(state).toBe(value);
+    });
```
I think the `setState` should be wrapped in `act` like you've done in the other tests and then do the assertions after it to better mimic how it actually runs? i.e. something like the following should work?

```suggestion
    const { result } = renderHook(() => usePersistedState("key", ""));
    const [state, setState] = result.current;
    expect(state).toBe("");
    act(() => setState(value));
    expect(state).toBe(value);
```

lucaslcode replied:
This was a brain fart

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+  });
+
+  it("saves then loads a value", () => {
+    const value = "Test string";
```
Might be worth testing this with an object value instead to check the stringifying and back works?

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+    expect(result.current.authState.authenticated).toBe(true);
+    await act(() => result.current.authActions.logout());
+    expect(result.current.authState.authenticated).toBe(false);
+    expect(result.current.authState.error).toBeNull();
```
And check the user is cleared too?

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+    try {
+      await act(() => result.current.authActions.updateJailStatus());
+    } catch (e) {
+      expect(e).toBeDefined();
```
I think it would be useful to check for the actual error message here instead so we know it's not some other random error being thrown.

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+    const state = result.current.authState;
+
+    // Things that have been updated are being reflected
+    expect(state.user).toMatchObject({
+      city: "New York",
+      hostingStatus: 3,
+      languages: ["English", "Finnish", "Spanish"],
+      countriesLivedList: ["Australia", "Finland", "Sweden", "United States"],
+      countriesVisitedList: ["Australia", "United States"],
+    });
```
```suggestion
    expect(state.user).toMatchObject({
      city: "New York",
      hostingStatus: 3,
      languages: ["English", "Finnish", "Spanish"],
      countriesLivedList: ["Australia", "Finland", "Sweden", "United States"],
      countriesVisitedList: ["Australia", "United States"],
      lat: 40.7306,
      lng: -73.9352
    });
```
Should check the new location properties too

darrenvong commented on app/frontend/src/features/auth/useAuthStore.test.ts:
```diff
@@ -0,0 +1,368 @@
+      countriesVisitedList: ["Australia", "United States"],
+    });
+    // Things haven't been updated should remain the same
+    expect(state.user).toMatchObject({
+      name: "Aapeli Vuorinen",
+      gender: "Male",
+      occupation: "Mathematician",
+      aboutMe: "Some generic stuff.",
+      aboutPlace: "About Aapeli's place",
+    });
```
```suggestion
    expect(state.user).toMatchObject({
      name: "Aapeli Vuorinen",
      gender: "Male",
      occupation: "Mathematician",
      aboutMe: "Some generic stuff.",
      aboutPlace: "About Aapeli's place",
      radius: 200
    });
```
Same reason as above.

lucaslcode committed:
Address review, add auth error handler test

lucaslcode committed:
review

lucaslcode committed:
Don't pass unnecessary error to handler

darrenvong commented on app/frontend/src/features/auth/login/UsernameForm.tsx:
```diff
@@ -57,7 +57,7 @@ export default function UsernameForm() {
+      authActions.authError(e.message);
     }
     setLoading(false);
   });
```
The setState on unmounted error is still there, and the safeState actually doesn't do anything. After some more digging... it looks like the **complete** conditional render below is the culprit so I suspect it's complaining about setState as the rendered component gets unmounted when they switch.

Inlining the `TextBody` stuff with `sent && <TextBody>...` so it's all in one return statement fixed it, but then that means you'll have to do another check to make sure the form only renders when it's `!sent`...

lucaslcode replied:
Fixed but I must admit I don't understand why the conditional render was causing this...

darrenvong replied:
Yeah it's a bit odd, since the component rendering it still exists. It's as if that unmounting means the `<TextBody>` gets a new copy of state all over again? 🤷

lucaslcode committed:
Remove conditional render, memoise setPersistedState

lucaslcode committed:
remove stray 'e' parameter

darrenvong commented:
Not sure if you want to leave this for another bug ticket, but here's what I found related to the question I asked on Slack earlier: if you sign up, but only accept one of the jail tasks then log out, the next time you log back in, you get an "invalid token" error banner in jail, even though you've successfully re-authenticated again.

And when you accepted all the jail tasks, while it correctly redirects you to the authenticated pages, the authenticated stuff doesn't work (e.g. if you try to update your profile, the page is blank) without having to log out and log back in again.

Quick demo video to show what I wrote in text (the "New" badge in GitHub is tempting me to upload one lol)

https://user-images.githubusercontent.com/13669362/103117037-970dc980-4660-11eb-95df-c575ef1bd4fc.mov

lucaslcode commented:
Okay solved it, the problem was this bit: 

```
useEffect(() => {
    if (urlToken && location.pathname !== loginPasswordRoute) {
      authActions.tokenLogin(urlToken);
    }
  }, [urlToken, authActions, location.pathname]);
```

authActions is memoized but depends on `authenticated`, so it gets a new actions object and sends again. Then, the token invalid error gives an UNAUTHORIZED code and logs out (in the newer branch) when it really should be NOT_FOUND.

I removed the dependency from authActions and it worked again. will commit

lucaslcode committed:
Remove authenticated dependence

lucaslcode committed:
Remove old test

darrenvong commented:
The user redirect after you accepted all the jail items are still broken in the latest preview :/ As in, you still have to log out and log back in again before you can perform an action to your profile (like trying to edit your profile)

lucaslcode committed:
set current user when leaving jail

lucaslcode committed:
Add AuthInterceptor tests

lucaslcode committed:
Merge pull request #452 from Couchers-org/frontend/bug/logout-when-unauthenticated

logout on unauthenticated error

lucaslcode commented:
Sorry merged other branch into here

lucaslcode committed:
Eliminate test warning about undefined user
